kytos/core/interface.py

```diff
@@ -173,19 +193,311 @@ def status_reason(self):
+            return index - 1
+
+        return None
+
+    def remove_tags(self, tags: list[int], tag_type: str = 'vlan') -> bool:
+        """Remove tags by resize available_tags
+        Returns False if nothing was remove, True otherwise"""
+        available = self.available_tags[tag_type]
+        index = self.find_index_remove(available, tags)
+        if index is None:
+            return False
+        # Resizing
+        if tags[0] == available[index][0]:
+            if tags[1] == available[index][1]:
+                available.pop(index)
+            else:
+                available[index: index+1] = [[tags[1]+1, available[index][1]]]
+        elif tags[1] == available[index][1]:
+            available[index: index+1] = [[available[index][0], tags[0]-1]]
+        else:
+            available[index: index+1] = [
+                [available[index][0], tags[0]-1],
+                [tags[1]+1, available[index][1]]
+            ]
+        return True
+
+    def use_tags(
+        self,
+        tags: list[int],
+        tag_type: str = 'vlan',
+        use_lock: bool = True,
+    ) -> bool:
+        """Remove a specific tag from available_tags if it is there.
+        Return False in case the tags were not able to be removed.
+        """
+        if use_lock:
+            with self._tag_lock:
+                result = self.remove_tags(tags, tag_type)
+        else:
+            result = self.remove_tags(tags, tag_type)
+        return result
+
+    @staticmethod
+    def find_index_add(
+        available_tags: list[list[int]],
+        tag_range: list[int]
+    ) -> Optional[int]:
+        """Find the index of tags in available_tags to be added"""
+        index = bisect.bisect_left(available_tags, tag_range)
+        previous, after = False, False
+        if index - 1 > -1:
+            if tag_range[0] > available_tags[index-1][1]:
+                previous = True
+        # Out of range
+        else:
+            previous = True
+
+        if index < len(available_tags):
+            if tag_range[1] < available_tags[index][0]:
+                after = True
+        else:
+            after = True
+
+        if previous and after:
+            return index
+        return None
+
+    # pylint: disable=too-many-branches
+    def add_tags(self, tags: list[int], tag_type: str = 'vlan') -> bool:
+        """Add tags, return True if they were added.
+        Returns False nothing nothing was added, True otherwise
+        Ensuring that ranges are not unnecessarily divided
+        available_tag e.g [[7, 10], [20, 30], [78, 92], [100, 109], [189, 200]]
+        tags examples are in each if statement.
+        """
+        if not tags[0] or not tags[1]:
+            return False
+        tag_ranges = self.tag_ranges[tag_type]
+        if self.find_index_remove(tag_ranges, tags) is None:
+            return False
+        available = self.available_tags[tag_type]
+        index = self.find_index_add(available, tags)
```

Comment:
You've nicely added a guarantee with `self.find_index_remove)` ensuring it's valid with `tag_ranges`, so at this point here whatever index that `self.find_index_add` return you'll need to add. With that said, there's room to make `self.find_index_add(available, tags)` simpler too, maybe just the result of `self.bisect_left` or even remove it, and then document in the docstring where it's supposed to be called and upfront validation like you've done with `find_index_remove(tag_ranges, tags)` before calling it. I'll encourage you to review this part and see if you can simplify it.

Reply:
`self.find_index_remove()` in this method can be deleted. To set a new `self.tag_range` every used TAG was taken into account already. So, if everything else is working well, any TAG to be added **should** be within the range of `self.tag_range` already. But could not help myself into not making sure that TAG can really be added.
Simplified `self.find_index_add()` so it is only one large if statement. But the result of `bisect_left` is also a suggested index since it does not look up for overlapping items.